### ISSUE TITLE
[Android][iOS] Support of stalling (or starving) events on mobile

### ISSUE THREAD

**Issue by midiphony** [RenderHeads/UnityPlugin-AVProVideo#872]
We would like to use the "Stalled"/"Unstalled" events on Android and iOS, but there are currently only supported on Windows.

Our use case : when the AVPro player is getting starved (video cannot be buffered), we would like to be notified in order to pause other systems.
_Feel free to share your thoughts or workarounds to handle this use case, with or without the support of the requested events._

Is it currently implementable ? (does ExoPlayer/AVFoundation provides that ?)

Thanks

**Comment by AndrewRH** [comment]
Which version of AVPro Video are you using?

**Comment by midiphony** [comment]
2.1.6

**Comment by AndrewRH** [comment]
I believe that should be working.  When the player isn't able to display any new frames, it enters a stalled state.

What media are you testing with?

**Comment by RenderHeadsSte** [comment]
I have checked on Android and _mediaPlayer.info.bool IsPlaybackStalled() returns true when a video loading off a server is having to buffer.

I believe you are after a stalled reporting when 'connection is dropped' ?  I believe you get Stalled reported on Windows and not Android when connection to server is chopped.

Can you please confirm this is the use-case/issue @midiphony ?

**Comment by midiphony** [comment]
@AndrewRH 
Hum, sorry, I haven't actually tested it !

I saw this comment in the declaration of AVPro events that paused my investigations : 
```
Stalled,			// Triggered when media is stalled (eg. when lost connection to media stream) - Currently only supported on Windows platforms
```

After reading both of your comments, I'm now realizing there's a default implementation for IsPlaybackStalled() that should work for all platforms.

Will the "Stalled"/"Unstalled" events be fired on all platforms or do we have to periodically check IsPlaybackStalled() ?


@RenderHeadsSte 
Our use case is actually the situation you first mentioned (online-hosted video having to buffer).
(but you are very right that I tried to trigger the event by chopping network connection)

**Comment by midiphony** [comment]
Nevermind, Stalled/Unstalled events + IsPlaybackStalled are indeed properly working on Android.
I tested on this HLS streaming : https://multiplatform-f.akamaihd.net/i/multi/april11/sintel/sintel-hd_,512x288_450_b,640x360_700_b,768x432_1000_b,1024x576_1400_m,.mp4.csmil/master.m3u8

When limiting/unblocking the carrier network bandwith, the events and states were as expected.

Very sorry for the inconvenience.

**I have been mistaken because of this situation :
When disabling all network access (carrier and WiFi) : video eventually stops playing, no stalling event is fired and IsPlaybackStalled doesn't return true. Playback never resumes when enabling again the network accesses.**
I guess there is an internal ExoPlayer error ("ExoPlayerImplInternal:   UnknownHostException (no network)") and I should handle the error properly, to potentially reload the video ?

(I can open another issue for clarity, and sorry again for the inconvenience.)

**Comment by RenderHeadsSte** [comment]
Yes, please do open a new issue :) 

I think we do have an issue here on Android, that connection chopping does not cause a stall event/situation.  I believe both Windows/macOS/iOS do this already. I will close this issue and make a new one specific to that.

**Comment by dr-formalyst** [comment]
For me (using 2.6.5) Stall/Unstall event will never fire on Android when using HLS streams from Vimeo. They do fire often when testing on Windows/Mac if I seek the video way ahead or back. I wait to play the video after FInishedSeeking event, and often the Stalled event will follow, video will be stalled for anywhere between 0.2 to sometimes a couple of seconds somtimes. I need to sync things to the video, and I can do that on Win/Mac keeping track of Stalled/Unstalled event but on Android (Quest 2) I can't do that because these event never fire. What may be the problem? Thanks :)

**Comment by dr-formalyst** [comment]
Oh i see that you mentioned that you will open a new issue specific to this -  can you please point to the new issue thread I can't find it?
